### PR TITLE
Simple dds

### PR DESCRIPTION
**Pull request by masathayde**
Ainda é necessário vários testes e correções, mas o básico parece funcionar.

### CODE REVIEW

File: simulator/driver.py [hunk 1]
```diff
@@ -3,10 +3,12 @@
 import simpy
 import logging
 
+import custom_error
```
Comment (gabrielsr) [inline simulator/driver.py]: Codacy Issue found: [Unused import custom_error](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=5588062)

**Comment by gabrielsr** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 17
           

Complexity increasing per file
==============================
- simulator/simple_dds/data_object.py  1
- simulator/simple_dds/entity.py  2
- simulator/simple_dds/topic.py  2
- simulator/simple_dds/domain_participant.py  2
- simulator/simple_dds/dds_service.py  4
- simulator/mockup.py  1
- simulator/simple_dds/subscriber.py  2
- simulator/singleton.py  2
- simulator/simple_dds/publisher.py  1
- simulator/custom_error.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=5588062&bid=18311621)

File: simulator/mockup.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
+import simpy
+import logging
```
Comment (gabrielsr) [inline simulator/mockup.py]: Codacy Issue found: [Unused import logging](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=5588062)

File: simulator/simple_dds/__init__.py [hunk 3]
```diff
@@ -0,0 +1,2 @@
+__all__ = ["entity", "dds_service", "domain_participant", "publisher", "subscriber",
```
Comment (gabrielsr) [inline simulator/simple_dds/__init__.py:1]: Codacy Issue found: [Undefined variable name 'entity' in __all__](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=5588062)

File: simulator/network.py [hunk 4]
```diff
@@ -2,6 +2,8 @@
 import simpy
 import logging
 
+import custom_error
```
Comment (gabrielsr) [inline simulator/network.py]: Codacy Issue found: [Unused import custom_error](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=5588062)

File: simulator/simple_dds/domain_participant.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        return new_subscriber
+
+    def delete_subscriber(self, subscriber):
+        pass
+
+    def get_discovered_participants(self):
+        pass
+        # Usar o service
+
+    def update_subscriber(self, subscriber, data_object):
```
Comment (gabrielsr) [inline simulator/simple_dds/domain_participant.py]: Codacy Issue found: [Method could be a function](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=5588062)

File: simulator/simple_dds/dds_service.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+        msg = ('ALL_DATA', local_data)
+        self.driver.async_function_call(['send', to_address, msg])
+    
+    def receive_full_domain_data(self, r_data):
+        # TODO: Remove this hack.
+        # HACK ALERT I JUST WANT THIS TO WORK
+        for element in r_data:
+            envelope = [0,0, element]
+            for z in self.interpret_data(envelope):
+                z
```
Comment (gabrielsr) [inline simulator/simple_dds/dds_service.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=5588062)

File: simulator/simple_dds/dds_service.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+from singleton import Singleton
+from simple_dds import entity
+
+# Possíveis problemas de uso mútuo nesta classe? Veremos quando testarmos.
+class UniqueHandleController(metaclass=Singleton):
+
+    def __init__(self):
+        self.next_available_handle = 1
+        self.lock = Lock()
+
```
Comment (gabrielsr) [inline simulator/simple_dds/dds_service.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=5588062)

File: simulator/simple_dds/entity.py [hunk 8]
```diff
@@ -0,0 +1,15 @@
+class Entity:
+
+    def __init__(self):
+        self.instance_handle = 0
+
```
Comment (gabrielsr) [inline simulator/simple_dds/entity.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=5588062)

File: simulator/simple_dds/dds_service.py [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+            participant.update_all_subscribers(data_object)
+
+    def attach_data_object_to_topic(self, data_object):
+        topic_name = data_object.get_topic_name()
+        if self.topic_exists(topic_name):
+            self.topics[topic_name].attach_data_object(data_object)
+
+    def send_full_domain_data(self, to_address):
+        local_data = []
+        for key, value in self.participants.items():
```
Comment (gabrielsr) [inline simulator/simple_dds/dds_service.py]: Codacy Issue found: [Unused variable 'key'](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=5588062)

File: simulator/network.py [hunk 10]
```diff
@@ -112,6 +116,13 @@ def end_lease(self, index):
         self.full_capacity = False
         logging.info(str(self.env.now) + ' :: ' + 'Lease ended for address {}'.format(address))
 
+    def send_addresses(self, driver):
+        addr_list = []
+        for peer in self.node_list:
+            if peer['address'] is not driver.address:
+                addr_list.append(peer['address'])
+        return addr_list
+
```
Comment (gabrielsr) [inline simulator/network.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=5588062)

File: simulator/mockup.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+# ... o uso do API simplifica bastante o trabalho do usuário, que não precisa se preocupar com o funcionamento da rede e detalhes de transmissão.
+# Só pra enfatizar, isto é uma simplificação EXTREMA do DDS, mais precisamente, da implementação OpenDDS. Eu tentei capturar o que eu compreendi sobre ele.
+
+class ApplicationMockup:
+
+    def __init__(self, domainID):
+        domainParticipantFactory = customdds.DomainParticipantFactoryImpl.get_instance()         # Singleton
+        # Criamos participante escolhendo uma ID para o domínio
+        self.participant = domainParticipantFactory.create_participant(domainID)
+
```
Comment (gabrielsr) [inline simulator/mockup.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=5588062)

File: simulator/mockup.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
+import simpy
```
Comment (gabrielsr) [inline simulator/mockup.py]: Codacy Issue found: [Unused import simpy](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=5588062)

File: simulator/driver.py [hunk 13]
```diff
@@ -18,40 +20,54 @@ def __init__(self, network, processor):
...
+                break     # Se chegarmos aqui, código completado com sucesso, saímos do loop
+            except RegistrationError as err:
+                print(err.message)
+                yield self.env.timeout(1)
+
+    def fetch_peer_list(self):
+        return self.network.send_addresses(self)
+
+    def disconnect(self):
+        former_address = self.address
```
Comment (gabrielsr) [inline simulator/driver.py]: Codacy Issue found: [Unused variable 'former_address'](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=5588062)

File: simulator/simple_dds/subscriber.py [hunk 14]
```diff
@@ -0,0 +1,30 @@
+import logging
+from queue import *
+
+from simple_dds import entity
+
+# TODO: Adicionar suporte para chamada de listener (callback).
+class Subscriber(entity.Entity):
+
```
Comment (gabrielsr) [inline simulator/simple_dds/subscriber.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=5588062)

File: simulator/simple_dds/dds_service.py [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+        self.topics = {}  # Topic name: Topic
+        self.data_objects = {} # Handle: Data object
+        self.message_handlers = {}
+
+        self.add_message_handler_methods()
+        self.attach_msg_reception_handler_to_driver()
+        self.discover_peers()
+        self.request_full_domain_data()
+
+    def set_instance_handle(self, handle):
```
Comment (gabrielsr) [inline simulator/simple_dds/dds_service.py:42]: Codacy Issue found: [Method could be a function](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=5588062)

File: simulator/simple_dds/data_object.py [hunk 16]
```diff
@@ -0,0 +1,15 @@
+from simple_dds import entity
+
+class Data_Object(entity.Entity):
+
```
Comment (gabrielsr) [inline simulator/simple_dds/data_object.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=5588062)

File: simulator/simple_dds/__init__.py [hunk 17]
```diff
@@ -0,0 +1,2 @@
...
+"data_object", "topic"]
```
Comment (gabrielsr) [inline simulator/simple_dds/__init__.py:2]: Codacy Issue found: [Undefined variable name 'data_object' in __all__](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=5588062)

File: simulator/driver.py (hunk 1 above)
Comment (masathayde) [reply simulator/driver.py]: Corrected.

File: simulator/simple_dds/data_object.py (hunk 16 above)
Comment (masathayde) [reply simulator/simple_dds/data_object.py]: Corrected.

File: simulator/simple_dds/subscriber.py (hunk 14 above)
Comment (masathayde) [reply simulator/simple_dds/subscriber.py]: Corrected.

File: simulator/driver.py (hunk 13 above)
Comment (masathayde) [reply simulator/driver.py]: Corrected.

File: simulator/mockup.py (hunk 12 above)
Comment (masathayde) [reply simulator/mockup.py]: Corrected.

File: simulator/mockup.py (hunk 11 above)
Comment (masathayde) [reply simulator/mockup.py]: Corrected.

File: simulator/network.py (hunk 10 above)
Comment (masathayde) [reply simulator/network.py]: Corrected.

File: simulator/simple_dds/dds_service.py (hunk 9 above)
Comment (masathayde) [reply simulator/simple_dds/dds_service.py]: Corrected.

File: simulator/simple_dds/entity.py (hunk 8 above)
Comment (masathayde) [reply simulator/simple_dds/entity.py]: Corrected.

File: simulator/simple_dds/dds_service.py (hunk 7 above)
Comment (masathayde) [reply simulator/simple_dds/dds_service.py]: Corrected.

File: simulator/simple_dds/dds_service.py (hunk 6 above)
Comment (masathayde) [reply simulator/simple_dds/dds_service.py]: Corrected.

File: simulator/simple_dds/domain_participant.py (hunk 5 above)
Comment (masathayde) [reply simulator/simple_dds/domain_participant.py]: Corrected.

File: simulator/network.py (hunk 4 above)
Comment (masathayde) [reply simulator/network.py]: Corrected.

File: simulator/mockup.py (hunk 2 above)
Comment (masathayde) [reply simulator/mockup.py]: Corrected.

File: simulator/simple_dds/__init__.py (hunk 3 above)
Comment (masathayde) [reply simulator/simple_dds/__init__.py:1]: Variable is needed for importing submodules contained in the folder.

File: simulator/simple_dds/dds_service.py (hunk 15 above)
Comment (masathayde) [reply simulator/simple_dds/dds_service.py:42]: Not in this case. This is an implementation of an abstract method, and, for this class, it's not supposed to interact or alter attributes. Instead, it raises an error, as we don't want it to ever be called.

File: simulator/simple_dds/__init__.py (hunk 17 above)
Comment (masathayde) [reply simulator/simple_dds/__init__.py:2]: Variable is needed for importing submodules contained in the folder.

**Comment by masathayde** [comment]
Ainda faltam métodos a completar, mas vou fazer o merge, por já funcionar de forma básica.

**Pull request merged**